Library/Homebrew/cmd/casks.sh

```diff
@@ -13,8 +13,11 @@ homebrew-casks() {
   if [[ -z "${HOMEBREW_NO_INSTALL_FROM_API}" &&
         -f "${HOMEBREW_CACHE}/api/cask_names.txt" ]]
   then
-    cat "${HOMEBREW_CACHE}/api/cask_names.txt"
-    echo
+    {
+      cat "${HOMEBREW_CACHE}/api/cask_names.txt"
+      echo
+      homebrew-items '*/Casks/*\.rb' '' 's|/Casks/|/|' '^homebrew/cask'
```

Comment:
I believe that we should exclude the main cask repo from `homebrew-items` output when using the API because if the local tap is out-of-date it might show some casks that have been deleted. This goes for `formula.sh` as well.

`homebrew-items()` is defined in `Library/Homebrew/items.sh`.

https://github.com/Homebrew/brew/blob/c265d3d330adb2707502e61cc93864fb9353a7bc/Library/Homebrew/items.sh#L22-L24

I wonder if we could just change `find_exclude_filter` to evaluate it as a regex instead of a name. Then, ignore `homebrew-cask` here and `homebrew-formula` in `formula.sh`. I haven't tried that though to see if it works.

Reply:
It does work per my testing but will changing `items.sh` cause breakage elsewhere?

Reply:
`items.sh` only defines the `homebrew-items` function which is only used in `casks.sh` and `formulae.sh` so it should be fine. Of course, we should double check that the output is the same when not using the API but that shouldn't be too difficult.

The idea of using `--regex` there to pass in multiple filter values doesn't really work as expected when I tried it partially because it doesn't accept empty values.

The best thing I can think of off the top of my head is something like this which should work but is quite ugly (I haven't tested this though).

<details>
<summary>Changes</summary>

```diff
diff --git a/Library/Homebrew/cmd/casks.sh b/Library/Homebrew/cmd/casks.sh
index dfa9308e6..98db12e2d 100644
--- a/Library/Homebrew/cmd/casks.sh
+++ b/Library/Homebrew/cmd/casks.sh
@@ -16,9 +16,9 @@ homebrew-casks() {
     {
       cat "${HOMEBREW_CACHE}/api/cask_names.txt"
       echo
-      homebrew-items '*/Casks/*\.rb' '' 's|/Casks/|/|' '^homebrew/cask'
+      homebrew-items '*/Casks/*\.rb' '' '*/homebrew-cask/*' 's|/Casks/|/|'
     } | sort -u
   else
-    homebrew-items '*/Casks/*\.rb' '' 's|/Casks/|/|' '^homebrew/cask'
+    homebrew-items '*/Casks/*\.rb' '' '' 's|/Casks/|/|' '^homebrew/cask'
   fi
 }
diff --git a/Library/Homebrew/cmd/formulae.sh b/Library/Homebrew/cmd/formulae.sh
index babbff33e..f4ee7e11f 100644
--- a/Library/Homebrew/cmd/formulae.sh
+++ b/Library/Homebrew/cmd/formulae.sh
@@ -16,10 +16,10 @@ homebrew-formulae() {
     {
       cat "${HOMEBREW_CACHE}/api/formula_names.txt"
       echo
-      homebrew-items '*\.rb' 'Casks' 's|/Formula/|/|' '^homebrew/core'
+      homebrew-items '*\.rb' 'Casks' '*/homebrew-core/*' 's|/Formula/|/|'
     } | sort -u
     echo
   else
-    homebrew-items '*\.rb' 'Casks' 's|/Formula/|/|' '^homebrew/core'
+    homebrew-items '*\.rb' 'Casks' '' 's|/Formula/|/|' '^homebrew/core'
   fi
 }
diff --git a/Library/Homebrew/items.sh b/Library/Homebrew/items.sh
index 4c436fed3..3c2ad8f21 100644
--- a/Library/Homebrew/items.sh
+++ b/Library/Homebrew/items.sh
@@ -1,10 +1,11 @@
 homebrew-items() {
   local items
   local sed_extended_regex_flag
-  local find_include_filter="$1"
-  local find_exclude_filter="$2"
-  local sed_filter="$3"
-  local grep_filter="$4"
+  local find_path_include_filter="$1"
+  local find_name_exclude_filter="$2"
+  local find_path_exclude_filter="$3"
+  local sed_filter="$4"
+  local grep_filter="$5"
 
   # HOMEBREW_MACOS is set by brew.sh
   # shellcheck disable=SC2154
@@ -21,14 +22,15 @@ homebrew-items() {
   items="$(
     find "${HOMEBREW_REPOSITORY}/Library/Taps" \
       -type d \( \
-      -name "${find_exclude_filter}" -o \
+      -name "${find_name_exclude_filter}" -o \
       -name cmd -o \
       -name .github -o \
       -name lib -o \
       -name spec -o \
-      -name vendor \
+      -name vendor -o \
+      -path "${find_path_exclude_filter}" \
       \) \
-      -prune -false -o -path "${find_include_filter}" |
+      -prune -false -o -path "${find_path_include_filter}" |
       sed "${sed_extended_regex_flag}" \
         -e 's/\.rb//g' \
         -e 's_.*/Taps/(.*)/(home|linux)brew-_\1/_' \
```

</details>

If we don't care about total accuracy with completions, we could just skip this extra complexity. Otherwise, hopefully someone else can chime in with a better idea.

Reply:
Since `items.sh` is only used in `casks.sh` and `formulae.sh`, it should still be safe to use `--regex` by passing a value that never matches anything like `^\b$` instead of an empty one. It still isn't the cleanest solution but in my opinion it is a lot cleaner than adding multiple extra arguments to `items.sh`.

Reply:
Yep, I agree with you. That makes more sense to me too.